javascript/elements/updates_for_element.js

```diff
@@ -45,71 +45,105 @@ export default class UpdatesForElement extends SubscribingElement {
     }
   }
 
-  update (data) {
+  async update (data) {
     // first updates-for element in the DOM *at any given moment* updates all of the others
-    if (this.blocks[0] !== this) return
+    if (this.blocks[0].element !== this) return
 
     // hold a reference to the active element so that it can be restored after the morph
     ActiveElement.set(document.activeElement)
 
     // store all retrieved HTML in an object keyed by URL to minimize fetch calls
     this.html = {}
 
+    const uniqueUrls = [...new Set(this.blocks.map(block => block.url))]
+
+    await Promise.all(
+      uniqueUrls.map(async url => {
+        if (!this.html.hasOwnProperty(url)) {
+          const response = await graciouslyFetch(url, {
+            'X-Cable-Ready': 'update'
+          })
+          this.html[url] = await response.text()
+        }
+      })
+    )
+
     // track current block index for each URL; referred to as fragments
     this.index = {}
 
-    this.blocks.forEach(this.processBlock.bind(this, data))
+    this.blocks.forEach(block => {
+      // if the block's URL is not in the index, initialize it to 0; otherwise, increment it
+      this.index.hasOwnProperty(block.url)
+        ? this.index[block.url]++
+        : (this.index[block.url] = 0)
+
+      block.process(data, this.html, this.index)
+    })
   }
 
-  async processBlock (data, block) {
-    // memoize the block's URL to avoid unnecessary DOM interop
-    const blockURL = url(block)
+  get blocks () {
+    // memoize blocks to avoid unnecessary DOM traversal
+    if (!this._blocks)
```

Comment:
Will memoizing `blocks` cause issues if the contents of the page changes? For example, if a second block was morphed into the page without disturbing this element, it seems like the 2nd block would never get operated upon.

In version I made, before I pulled `blocks` out into a getter, I made a local copy of the current `blocks` snapshot inside of the `update` method. It was only intended to exist for the duration of the `update` call.

Now that we have an asynch `update`, we need to be really careful about what we're memoizing. I actually do think we need to revert this to something that exists only during the life of one `update` call.

Reply:
This was just refactored by me... I'm fine with de-memoizing it, but I think that should be it, then it's just a simple wrapper around `document.querySelectorAll`

Reply:
okay, I think now get you... let me just quickly fix this